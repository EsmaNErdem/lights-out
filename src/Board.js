import React from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 5 , ncols = 5, chanceLightStartsOn = 0.25 }) {
  const [board, setBoard] = React.useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    for (let i = 0; i < nrows; i++) {
      const col = Array.from({length: ncols}, () =>  Math.random() < chanceLightStartsOn);
      initialBoard.push(col);
    }
    return initialBoard;
  }

  function hasWon() {
    return board.every(r => r.every(c => c == false))
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      const copyBoard = board.map(row => [...row])

      flipCell(y, x, copyBoard)
      flipCell(y+1, x, copyBoard)
      flipCell(y-1, x, copyBoard)
      flipCell(y, x+1, copyBoard)
      flipCell(y, x-1, copyBoard)

      return copyBoard
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  if( hasWon() ) {
    return (
      <h1>You won the game!</h1>
    )
  }   

  return (
    <table className="Board">
      <tbody>
        {board.map((r, yCoord ) => ( 
          <tr key={yCoord}>
            {r.map((c, xCoord) => (
              <Cell
                key={`${yCoord}-${xCoord}`} 
                isLit={c}
                coord={`${yCoord}-${xCoord}`}
                flipCellsAroundMe={flipCellsAround}
              />
            ))} 
            {/* {} needed within tr tag */}
          </tr>
        ))}
      </tbody>
    </table>
  )
}

export default Board;
