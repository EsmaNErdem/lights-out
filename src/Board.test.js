import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

beforeEach(() => {
  // Mock Math.random to return a value less than 0.25
  jest.spyOn(Math, "random").mockReturnValue(0.10); 
});

afterEach(() => {
  jest.restoreAllMocks(); // Restore Math.random after each test. if it was in the end, it will restore after all tests are rendered
});

test("renders without crashing", () => {
  render(<Board />);
});

test("it renders with all cells lit", () => {
  const { asFragment } = render(<Board />);
  expect(asFragment()).toMatchSnapshot();
});

test("it renders with all cells unlit so the game is won", () => {
    // Mock Math.random to return a value more than 0.25
    jest.spyOn(Math, "random").mockReturnValue(0.8);

    const { asFragment } = render(<Board />);
    expect(asFragment()).toMatchSnapshot();

    // Restore the original Math.random after the test
    jest.spyOn(Math, "random").mockRestore();
});

test("it renders with all cells unlit so the game is won, check if the game show h1", () => {
    // Mock Math.random to return a value more than 0.25
    jest.spyOn(Math, "random").mockReturnValue(0.8);

    const { getByText } = render(<Board />);
    expect(getByText("You won the game!")).toBeInTheDocument();

    // Restore the original Math.random after the test
    jest.spyOn(Math, "random").mockRestore();
});

test("test clicking a cell will turn the lights on or off", () => {
    const { queryByTestId } = render(<Board />);

    const cell22 = queryByTestId("2-2")
   
    const cell21 = queryByTestId("2-1") // left
    const cell12 = queryByTestId("1-2") // top
    const cell23 = queryByTestId("2-3") // right
    const cell32 = queryByTestId("3-2") // bottom

    expect(cell22).toHaveClass("Cell-lit")
    expect(cell21).toHaveClass("Cell-lit")
    expect(cell12).toHaveClass("Cell-lit")
    expect(cell23).toHaveClass("Cell-lit")
    expect(cell32).toHaveClass("Cell-lit")

    // Click cells to simulate the game
    fireEvent.click(cell22)
    
    expect(cell22).not.toHaveClass("Cell-lit")
    expect(cell21).not.toHaveClass("Cell-lit")
    expect(cell12).not.toHaveClass("Cell-lit")
    expect(cell23).not.toHaveClass("Cell-lit")
    expect(cell32).not.toHaveClass("Cell-lit")
})

test("test clicking a cell will turn the lights on or off only on the neighboring cells", () => {
    const { queryByTestId } = render(<Board />);

    const cell22 = queryByTestId("2-2")
   
    const cell01 = queryByTestId("0-1") // random cell

    expect(cell22).toHaveClass("Cell-lit")
    expect(cell01).toHaveClass("Cell-lit")
    
    // Click cells to simulate the game
    fireEvent.click(cell22)
    
    expect(cell22).not.toHaveClass("Cell-lit")
    expect(cell01).toHaveClass("Cell-lit")
})

test("it renders a win when the winning condition is met", () => {

    // HOW CAN I MOCK BOARD'S STATE???????

    //custom board with a known winning condition
    const customBoard = [
        [false, true, false],
        [true, true, true],
        [false, true, false],
        ];

    const { container } = render(
        <Board nrows={3} ncols={3} chanceLightStartsOn={0} /> // Ensure lights won't start on
    );
  

    // SO I CAN TEST THIS:
  
    // //winning message is not displayed initially
    // expect(queryByText("You won the game!")).toBeNull();
  
    // // Click cells to simulate the game
    // fireEvent.click(queryByTestId("1-1")); 
  
    // // Assert that the game declares a win
    // expect(getByText("You won the game!")).toBeInTheDocument();
  });


