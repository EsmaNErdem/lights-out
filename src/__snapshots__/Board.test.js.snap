// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders with all cells lit 1`] = `
<DocumentFragment>
  <table
    class="Board"
  >
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="0-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="0-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="1-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="1-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="2-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="2-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="3-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="3-4"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          data-testid="4-0"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-1"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-2"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-3"
        />
        <td
          class="Cell Cell-lit"
          data-testid="4-4"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`it renders with all cells unlit so the game is won 1`] = `
<DocumentFragment>
  <h1>
    You won the game!
  </h1>
</DocumentFragment>
`;
